plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -73,230 +151,453 @@ their margin. If either party fails to make a margin payment then the contract
-settleEarly :: (
-    MonadError WalletAPIError m,
-    WalletAPI m)
-    => [TxOutRef]
-    -> Future
-    -> FutureData
-    -> OracleValue Ada
-    -> m ()
-settleEarly refs ft fd ov = do
-    let totalVal = Ada.toValue ((futureDataMarginLong fd) + (futureDataMarginShort fd))
-        outs = [Ledger.pubKeyTxOut totalVal (futureDataLong fd)]
-        inp = (\r -> scriptTxIn r (validatorScript ft) red dat) <$> refs
-        red = RedeemerScript $ PlutusTx.toData $ Settle ov
-        dat = DataScript $ PlutusTx.toData $ fd
-    void $ createTxAndSubmit defaultSlotRange (Set.fromList inp) outs []
-
-adjustMargin :: (
-    MonadError WalletAPIError m,
-    WalletAPI m)
-    => [TxOutRef]
-    -> Future
-    -> FutureData
-    -> Ada
-    -> m ()
-adjustMargin refs ft fd vl = do
-    pk <- ownPubKey
-    (payment, change) <- createPaymentWithChange (Ada.toValue vl)
-    fd' <- let fd''
-                | pk == futureDataLong fd = pure $ fd { futureDataMarginLong  = vl + futureDataMarginLong fd  }
-                | pk == futureDataShort fd = pure $ fd { futureDataMarginShort = vl + futureDataMarginShort fd }
-                | otherwise = throwOtherError "Private key is not part of futures contrat"
-            in fd''
-    let
-        red = RedeemerScript $ PlutusTx.toData AdjustMargin
-        dat  = DataScript $ PlutusTx.toData fd
-        dat'  = DataScript $ PlutusTx.toData fd'
-        o = scriptTxOut outVal (validatorScript ft) dat'
-        outVal = Ada.toValue (vl + futureDataMarginLong fd + futureDataMarginShort fd)
-        inp = Set.fromList $ (\r -> scriptTxIn r (validatorScript ft) red dat) <$> refs
-    void $ createTxAndSubmit defaultSlotRange (Set.union payment inp) (o : maybeToList change) [dat']
+data Future =
+    Future
+        { ftDeliveryDate  :: Slot
+        , ftUnits         :: Integer
+        , ftUnitPrice     :: Value
+        , ftInitialMargin :: Value
+        , ftPriceOracle   :: PubKey
+        , ftMarginPenalty :: Value
+        -- ^ How much a participant loses if they fail to make the required
+        --   margin payments.
+        } deriving Generic
+
+-- | The tokens that represent ownership of the two sides of the future
+data FutureOwners =
+    FutureOwners
+        { ftoLong  :: AccountOwner
+        -- ^ The owner of the "long" account (represented by a token)
+        , ftoLongAccount :: ValidatorHash
+        -- ^ Address of the 'TokenAccount' validator script for 'ftoLong'
+        , ftoShort :: AccountOwner
+        -- ^ The owner of the "short" account (represented by a token)
+        , ftoShortAccount :: ValidatorHash
+        -- ^ Address of the 'TokenAccount' validator script for 'ftoShort'
+        } deriving (Haskell.Show, Generic)
+
+mkOwners
+    :: AccountOwner
+    -> AccountOwner
+    -> FutureOwners
+mkOwners long short =
+    FutureOwners
+        { ftoLong = long
+        , ftoLongAccount = TokenAccount.validatorHash long
```

Comment:
Verily it is derivable. So why not just do this at the use site?

Reply:
We need this in the on-chain code, so it needs to be `lift`ed in

Reply:
This is the hash of the validator script of the token account applied to `long`

Reply:
Right, makes sense. Comment, though!